sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java

```diff
@@ -0,0 +1,181 @@
+      final ITransaction transaction =
+          hub.startTransaction(getActivityName(activity), "navigation");
+
+      // lets bind to the scope so other integrations can pick it up
+      hub.configureScope(
+          scope -> {
+            // we we'd not like to overwrite existent transactions bound to the Scope manually.
+            if (scope.getTransaction() == null) {
+              scope.setTransaction(transaction);
+            }
+          });
+
+      activities.put(activity, transaction);
+    }
+  }
+
+  private boolean isRunningTransaction(final @NonNull Activity activity) {
+    return activities.containsKey(activity);
+  }
+
+  private void stopTracing(final @NonNull Activity activity) {
+    if (performanceEnabled && options.isEnableAutoActivityLifecycleTracingFinish()) {
+      final ITransaction transaction = activities.get(activity);
+      if (transaction != null && !transaction.isFinished()) {
+        SpanStatus status = transaction.getStatus();
+        // status might be set by other integrations, let's not overwrite it
+        if (status == null) {
+          status = SpanStatus.OK;
+        }
+        transaction.finish(status);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void onActivityCreated(
+      @NonNull Activity activity, @Nullable Bundle savedInstanceState) {
+    addBreadcrumb(activity, "created");
+
+    // if activity has global fields being init. and
+    // they are slow, this won't count the whole fields/ctor initialization time, but only
+    // when onCreate is actually called.
+    startTracing(activity);
+  }
+
+  @Override
+  public synchronized void onActivityStarted(@NonNull Activity activity) {
+    addBreadcrumb(activity, "started");
+  }
+
+  @Override
+  public synchronized void onActivityResumed(@NonNull Activity activity) {
+    addBreadcrumb(activity, "resumed");
+  }
+
+  @Override
+  public synchronized void onActivityPostResumed(@NonNull Activity activity) {
+    // this should be called only when onResume has been executed already, which means
+    // the UI is responsive at this moment.
+    stopTracing(activity);
+  }
+
+  @Override
+  public synchronized void onActivityPaused(@NonNull Activity activity) {
+    addBreadcrumb(activity, "paused");
+  }
+
+  @Override
+  public synchronized void onActivityStopped(@NonNull Activity activity) {
+    addBreadcrumb(activity, "stopped");
+
+    // clear up so we dont start again for the same activity if the activity is in the acitvity
+    // stack still
+    // if the activity is opened again and not in memory, transactions will be created normally.
+    if (performanceEnabled) {
+      activities.remove(activity);
+    }
+  }
+
+  @Override
+  public synchronized void onActivitySaveInstanceState(
+      @NonNull Activity activity, @NonNull Bundle outState) {
+    addBreadcrumb(activity, "saveInstanceState");
+  }
+
+  @Override
+  public synchronized void onActivityDestroyed(@NonNull Activity activity) {
```

Comment:
Could this be called instead of `onActivityPostResumed` leaving an open transaction hanging?

Such edge case makes me realize we probably want to implement 1k (or other value) max Span and start dropping it because if we end up with a transaction that isnt' finished, we could add spans to it indefinitely. Another edge case is a transaction that's finished but still bound to the scope and the SDK allows adding spans to finished transactions. We need to disallow that and log out a warning.

Reply:
not sure if I got it, onActivityPostResumed means the App went foreground.
`onActivityDestroyed` means App went background.

if you meant `onActivityStopped` instead of `onActivityPostResumed`, the answer is no, `onActivityStopped` always run before `onActivityDestroyed` hence a transaction will never be hanging unless the user opt-out `enableAutoActivityLifecycleTracingFinish` and didn't finish the transaction manually and never closed the screen either, we could be sure that the transaction is ended at `onActivityDestroyed` even if `enableAutoActivityLifecycleTracingFinish` was disabled.

Reply:
I will end the transaction regardely if `enableAutoActivityLifecycleTracingFinish` is disabled, on `onActivityDestroyed`

Reply:
Makes sense, thanks for clarifying.